### PR TITLE
Support slicing edgedb.Tuple

### PR DESCRIPTION
**Pull request by fmoor**
Working towards https://github.com/edgedb/edgedb-python/issues/296

### CODE REVIEW

File: edgedb/datatypes/tuple.c [hunk 1]
```diff
@@ -172,6 +172,98 @@ tuple_getitem(EdgeTupleObject *o, Py_ssize_t i)
...
+        for (Py_ssize_t i = start; i < stop; i += step) {
+            new_size += 1;
+        }
+    }
+
+    PyObject *n = EdgeTuple_New(new_size);
+    Py_ssize_t j = 0;
+    if (step < 0) {
+        for (Py_ssize_t i = start; i > stop; i += step) {
+            PyObject *el = PyTuple_GET_ITEM(o, i + 1);
```
Comment (fmoor) [inline edgedb/datatypes/tuple.c:223]: Why is this `i + 1` and not `i`?

**Comment by fmoor** [comment]
I have never touched a c extension before :/

File: edgedb/datatypes/tuple.c [hunk 2]
```diff
@@ -172,6 +172,98 @@ tuple_getitem(EdgeTupleObject *o, Py_ssize_t i)
...
+
+        if (start == 0 && stop == size && step == 1) {
+            Py_INCREF(o);
+            return o;
+        }
+    }
+
+    Py_ssize_t new_size = 0;
+    if (step < 0) {
+        for (Py_ssize_t i = start; i > stop; i += step) {
```
Comment (fmoor) [inline edgedb/datatypes/tuple.c:210]: I'm pretty sure there is a better way than this to calculate the new tuple's size...

Comment (elprans) [reply edgedb/datatypes/tuple.c:210]: See how Python does this: https://github.com/python/cpython/blob/0a40025b803a8dc1d604538e317992827ab96625/Objects/tupleobject.c#L779

File: edgedb/datatypes/tuple.c [hunk 3]
```diff
@@ -172,6 +172,60 @@ tuple_getitem(EdgeTupleObject *o, Py_ssize_t i)
...
+    }
+
+    if (start == 0 && step == 1 && slicelength == size) {
+        Py_INCREF(o);
+        return o;
+    }
+
+    PyObject *result = EdgeTuple_New(slicelength);
+    for (cur = start, i = 0; i < slicelength; cur += step, i++) {
+        PyObject *el = PySequence_GetItem(o, cur);
```
Comment (elprans) [inline edgedb/datatypes/tuple.c:192]:
```suggestion
        PyObject *el = EdgeTuple_GET_ITEM(o, i);
```

File: edgedb/datatypes/tuple.c [hunk 4]
```diff
@@ -172,6 +172,60 @@ tuple_getitem(EdgeTupleObject *o, Py_ssize_t i)
...
+        return tuple_getslice(o, start, stop, step);
+    }
+
+    if (PyIndex_Check(key)) {
+        Py_ssize_t val = PyNumber_AsSsize_t(key, PyExc_IndexError);
+        if (val == -1 && PyErr_Occurred()) {
+            return NULL;
+        }
+
+        return PySequence_GetItem(o, val);
```
Comment (elprans) [inline edgedb/datatypes/tuple.c:219]:
```suggestion
        return EdgeTuple_GET_ITEM(o, val);
```

Comment (elprans) [inline edgedb/datatypes/tuple.c:219]:
```suggestion
        Py_ssize_t val = PyNumber_AsSsize_t(key, PyExc_IndexError);
        if (val == -1 && PyErr_Occurred()) {
            return NULL;
        }

        return tuple_getitem(o, val);
```

**Review by 1st1 — APPROVED** [review]

**Comment by fmoor** [comment]
> We'll still have to have `edgedb.NamedTuple` fwiw, but I don't see any support of new slicing capabilities added to it. Why?

I had never worked on a C extension before so I was starting small.

**Comment by 1st1** [comment]
> I had never worked on a C extension before so I was starting small.

NP, but since this is now done and looks good, I suggest to generalize the functions into `_EdgeGeneric_*` family and reuse them from both the namedtuple and tuple (because both types support slicing etc).

**Comment by 1st1** [comment]
Closing this one -- we got rid of edgedb.Tuple

**Pull request closed without merging**